src/wallet/wallet.cpp

```diff
@@ -426,27 +442,289 @@ ZcashdUnifiedSpendingKey CWallet::GenerateNewUnifiedSpendingKey() {
+        auto ufvkid = ufvk.value().GetKeyID();
+
+        // Check whether an address has already been generated for any provided
+        // diversifier index. Return that address, or set the diversifier index
+        // at which we'll begin searching for the next available diversified
+        // address.
+        auto metadata = mapUfvkAddressMetadata.find(ufvkid);
+        if (metadata != mapUfvkAddressMetadata.end()) {
+            if (j.has_value()) {
+
+                auto receivers = metadata->second.GetReceivers(j.value());
+                if (receivers.has_value()) {
+                    // Ensure that the set of receiver types being requested is
+                    // the same as the set of receiver types that was previously
+                    // generated. If they match, simply return that address.
+                    if (receivers.value() == receiverTypes) {
+                        return std::make_pair(ufvk.value().Address(j.value(), receiverTypes).value(), j.value());
+                    } else {
+                        return AddressGenerationError::ExistingAddressMismatch;
+                    }
+                }
+            } else {
+                // Set the diversifier index to one greater than the last used
+                // diversifier
+                j = metadata->second.GetNextDiversifierIndex();
+                if (!j.has_value()) {
+                    return AddressGenerationError::DiversifierSpaceExhausted;
+                }
+            }
+        } else {
+            // Begin searching from the zero diversifier index if we haven't
+            // yet generated an address from the specified UFVK and no
+            // diversifier index has been specified.
+            if (!j.has_value()) {
+                j = libzcash::diversifier_index_t(0);
+            }
+        }
+
+        // Find a working diversifier and construct the associated address.
+        // At this point, we know that `j` will contain a value. Optimistically
+        // attempt to find an address at that diversifier; we'll search if we
+        // don't find one.
+        auto diversifierIndex = j.value();
+        auto address = ufvk.value().Address(diversifierIndex, receiverTypes);
+        if (!address.has_value() && searchDiversifiers) {
+            auto found = ufvk.value().FindAddress(j.value(), receiverTypes);
+            if (found.has_value()) {
+                address = found.value().first;
+                diversifierIndex = found.value().second;
+            } else {
+                return AddressGenerationError::DiversifierSpaceExhausted;
+            }
+        }
+
+        // Now that we're done searching, check that we actually got an address
+        if (!address.has_value()) {
+            return AddressGenerationError::NoAddressForDiversifier;
+        }
+
+        // Persist the newly created address to the keystore
+        mapUfvkAddressMetadata[ufvkid].SetReceivers(diversifierIndex, receiverTypes);
```

Comment:
The return value isn't being checked. We should either be checking it and returning an error if false, or just asserting that it returns true.